convert-mpt-hf-to-gguf.py

```diff
@@ -0,0 +1,264 @@
+    parser.add_argument(
+        "--vocab-only", action="store_true",
+        help="extract only the vocab",
+    )
+    parser.add_argument(
+        "--outfile", type=Path,
+        help="path to write to; default: based on input",
+    )
+    parser.add_argument(
+        "model", type=Path,
+        help="directory containing model file, or model file itself (*.bin)",
+    )
+    parser.add_argument(
+        "ftype", type=int, choices=[0, 1], default=1, nargs='?',
+        help="output format - use 0 for float32, 1 for float16",
+    )
+    return parser.parse_args()
+
+args = parse_args()
+
+dir_model = args.model
+ftype = args.ftype
+if not dir_model.is_dir():
+    print(f'Error: {args.model} is not a directory', file = sys.stderr)
+    sys.exit(1)
+
+# possible tensor data types
+#   ftype == 0 -> float32
+#   ftype == 1 -> float16
+
+# map from ftype to string
+ftype_str = ["f32", "f16"]
+
+if args.outfile is not None:
+    fname_out = args.outfile
+else:
+    # output in the same directory as the model by default
+    fname_out = dir_model / f'ggml-model-{ftype_str[ftype]}.gguf'
+
+print("gguf: loading model "+dir_model.name)
+
+with open(dir_model / "config.json", "r", encoding="utf-8") as f:
+    hparams = json.load(f)
+
+if hparams["architectures"][0] != "MPTForCausalLM":
+    print("Model architecture not supported: " + hparams["architectures"][0])
+
+    sys.exit()
+
+# get number of model parts
+num_parts = count_model_parts(dir_model)
+
+ARCH=gguf.MODEL_ARCH.MPT
+gguf_writer = gguf.GGUFWriter(fname_out, gguf.MODEL_ARCH_NAMES[ARCH])
+
+print("gguf: get model metadata")
+
+block_count = hparams["n_layers"]
+
+gguf_writer.add_name(dir_model.name)
+gguf_writer.add_context_length(hparams["max_seq_len"])
+gguf_writer.add_embedding_length(hparams["d_model"])
+gguf_writer.add_block_count(block_count)
+gguf_writer.add_feed_forward_length(4 * hparams["d_model"])
+gguf_writer.add_head_count(hparams["n_heads"])
+gguf_writer.add_layer_norm_eps(1e-05) 
+if hparams["attn_config"]["clip_qkv"] is not None:
+    gguf_writer.add_clamp_kqv(hparams["attn_config"]["clip_qkv"])
+gguf_writer.add_max_alibi_bias(hparams["attn_config"]["alibi_bias_max"])
+
+# TOKENIZATION
+
+print("gguf: get tokenizer metadata")
+
+tokens: list[bytearray] = []
+
+tokenizer_json_file = dir_model / 'tokenizer.json'
+if not tokenizer_json_file.is_file():
+    print(f'Error: Missing {tokenizer_json_file}', file = sys.stderr)
+    sys.exit(1)
+
+# gpt2 tokenizer
+gguf_writer.add_tokenizer_model("gpt2")
+
+with open(tokenizer_json_file, "r", encoding="utf-8") as f:
+    tokenizer_json = json.load(f)
+
+print("gguf: get gpt2 tokenizer vocab")
+
+# MPT token embedding tensors have dimension 50432, but there are only 50254
+# tokens in the vocab, presumably to accomodate some "reserved" tokens;
+# this is causing problems down the line in llama.cpp, so we extend the vocab_size:
+
+vocab_size = len(tokenizer_json["model"]["vocab"]) + 178
```

Comment:
You inherited a few bugs from the GPT-NeoX script, and then you hardcoded a constant in order to work around one of them. This explains why you didn't come across https://github.com/ggerganov/llama.cpp/pull/3021#issuecomment-1738041891 like I did a few days ago while writing an MPT conversion script - I believe I based my code on the Falcon script.

See #3421 and #3405, but as with this PR, they currently conflict with the PR with the BPE tokenizer changes.

Reply:
I did come across this KeyError / c vs. ord(c) bug and built in a primitive workaround (should have mentioned this in the PR comment, too, but forgot). I may have erroneously assumed that the discrepancy between vocab_size and embedding tensor size is due to MPT padding the tensor to allow for future/custom vocab extensions, based on what I remembered from https://huggingface.co/mosaicml/mpt-7b:

> The model vocabulary size of 50432 was set to be a multiple of 128 (as in [MEGATRON-LM](https://arxiv.org/abs/1909.08053)), model flop utilization (MFU) increased by up to four percentage points.

I believe I checked that there were actually fewer than 50432 tokens (even with the "added tokens"). But overall, I haven't looked deeply enough into this issue, as the priority was to get the generation working in the first place.

Reply:
I just checked that the tokenization code (gptneox+cpp) from pr:3252 can be used for MPT, with two slight adaptations:
1. The vocab_size += 178 must stay (as this covers the dummy/non-existent tokens on top of the added_tokens)
2. As you mentioned in https://github.com/ggerganov/llama.cpp/pull/3252#issuecomment-1742181655, there would be a KeyError because of len(reverse_vocab) < vocab_size. It can be worked around by constructing a dummy token similar to what is already done in this PR:
```
-    tokens.append(text)
+    tokens.append(reverse_vocab[i] if i in reverse_vocab else f"[PAD{i}]")
```

Reply:
Thanks for checking. It should be straightforward for you to add a new test case for `test-tokenizer-1-bpe`. I'd be interested in the results.

Reply:
@goerch test-tokenizer-1-bpe currently fails for the MPT vocab:
```
main : info: utf8 conversion 49835 'ç¡'
main : info: utf8 conversion 50152 'ë '
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at

Program received signal SIGABRT, Aborted.
```
The utf8 argument of unicode_to_bytes_bpe which causes this failure is " " (a space) at i=50254. This is the first of the multi-space added_tokens from MPT's tokenizer.json.

Reply:
> This is the first of the multi-space added_tokens from MPT's tokenizer.json.

Interesting, thanks for checking this. I'll try to cover some models with added tokens with these tests by the end of the week.

Reply:
See #3405 for the fix that worked prior to #3252 being merged. I have not looked into that PR deep enough to understand what the new fix would look like.

Reply:
```
> terminate called after throwing an instance of 'std::out_of_range'
>   what():  _Map_base::at
```

I'm trying to reproduce and understand that breakage. I'm not sure I'm qualified to merge `main` into your branch though.

Reply:
I merged PR 3252 into my mpt branch instead, and updated convert-mpt-hf-to-gguf.py to match your convert-gptgeox-hf-to-gguf.py as much as possible. I pushed the result to https://github.com/jploski/llama.cpp/tree/mpt-pr-3252 so that you can reproduce the error in test-tokenizer-1-bpe while running on the converted MPT vocab.

Reply:
Leaving the `test-tokenizer-1-bpe` test aside, is the MPT implementation working as expected?
If so, should we merge and resolve the failing test later from `master`. What would you guys recommend?

Reply:
@jploski  : I'm sorry for delaying all of this due to #3502. But I think 

- I saw some small mistakes in `bpe_gpt2_preprocess` (insignificant, will post later)
- Regarding the most important differences in #3502 I'm totally unsure if this is a bug on our side or in HF's fast tokenizers

The problem AFAIU regards 4 digit dates like 1911 for example. We tokenize it (for the `Falcon` vocabulary) as 

```
660 '19'
1069 '11'
```

and HF tokenizes it to 

```
6474 '191'
28 '1'
```

and I *think* they are wrong. But how to argue about this or find the difference in algorithms is my current problem.